saleor/payment/gateways/braintree/__init__.py

```diff
@@ -136,128 +118,111 @@ def get_client_token(**connection_params: Dict) -> str:
     return client_token
 
 
+def process_payment(
+        payment: Dict, payment_token: str, amount: Decimal,
+        **connection_params: Dict) -> Dict:
+    auth_resp = authorize(payment, payment_token, **connection_params)
+    transaction_token = auth_resp['transaction_id']
+
+    try:
+        return capture(transaction_token, amount, **connection_params)
+    except Exception as e:
+        void(transaction_token, **connection_params)
+        raise
+
+
 def authorize(
-        payment: Payment,
-        payment_token: str,
-        **connection_params: Dict) -> Tuple[Transaction, str]:
+        payment: Dict, payment_token: str, **connection_params: Dict) -> Dict:
     gateway = get_braintree_gateway(**connection_params)
+
     try:
         result = gateway.transaction.sale({
-            'amount': str(payment.total),
+            'amount': str(payment['total']),
             'payment_method_nonce': payment_token,
             'options': {
                 'submit_for_settlement': CONFIRM_MANUALLY,
                 'three_d_secure': {
                     'required': THREE_D_SECURE_REQUIRED}},
             **get_customer_data(payment)})
     except braintree_sdk.exceptions.NotFoundError:
-        return transaction_and_incorrect_token_error(
-            payment, kind=TransactionKind.AUTH, token=payment_token)
+        # FIXME: make this a payment generic exception
+        raise Exception('Unable to process the transaction. '
+            'Transaction\'s token is incorrect or expired.')
+
     gateway_response = extract_gateway_response(result)
     error = get_error_for_client(gateway_response['errors'])
-    credit_card_data = gateway_response.pop('credit_cart', None)
-    if credit_card_data:
-        payment.cc_first_digits = credit_card_data.bin
-        payment.cc_last_digits = credit_card_data.last_4
-        payment.cc_brand = credit_card_data.card_type
-        payment.cc_exp_month = credit_card_data.expiration_month
-        payment.cc_exp_year = credit_card_data.expiration_year
-        payment.save(update_fields=[
-            'cc_first_digits', 'cc_last_digits', 'cc_brand',
-            'cc_exp_month', 'cc_exp_year'])
-
-    txn = create_transaction(
-        payment=payment,
-        kind=TransactionKind.AUTH,
-        amount=gateway_response.pop('amount', payment.total),
-        currency=gateway_response.pop('currency', payment.currency),
-        gateway_response=gateway_response,
-        token=getattr(result.transaction, 'id', ''),
-        is_success=result.is_success)
-    return txn, error
+
+    return {
+        'is_success': result.is_success,
+        'transaction_id': gateway_response['transaction_id'],
+        'raw_response': gateway_response,
```

Comment:
I don't agree, why make a response just for the sake of having a response?

Reply:
Because this is a little bit confusing. There should still be a raw response though it is a dummy gateway. You are wring a standard, if this is not kept as a convention, it requires docs explaining for that. So why not do it in code.

Reply:
The docs will be updated to match, and everything will be explained. Certain gateways will be provided as examples as well. Dummy.. I'd say is not a good example, it's basically a debug gateway to just force the payment into a certain state. Besides currently it's kinda broken :D

Reply:
That's will be great.